File: app/src/main/java/com/benhero/glstudio/base/GLAnimationSet.java
package com.benhero.glstudio.base;

import java.util.ArrayList;
import java.util.List;

/**
 * GL动画集合
 *
 * @author Benhero
 */
public class GLAnimationSet extends GLAnimation {
    private List<GLAnimation> mAnimations = new ArrayList<>();

    public void addAnimation(GLAnimation animation) {
        mAnimations.add(animation);
    }

    public List<GLAnimation> getAnimations() {
        return mAnimations;
    }

    @Override
    public void setStartTime(long startTime) {
        super.setStartTime(startTime);
        for (GLAnimation animation : mAnimations) {
            animation.setStartTime(startTime);
        }
    }

    @Override
    public void setDuration(long duration) {
        super.setDuration(duration);
        for (GLAnimation animation : mAnimations) {
            animation.setDuration(duration);
        }
    }

    /**
     * 根据子动画更新动画集合的时间
     */
    public void updateTimeByChildren() {
        long startTime = 0;
        long endTime = 0;
        for (GLAnimation animation : mAnimations) {
            startTime = Math.min(mStartTime, animation.mStartTime);
            endTime = Math.max(animation.getStartTime() + animation.getDuration(), endTime);
        }
        mStartTime = startTime;
        mDuration = endTime - startTime;
    }

    @Override
    public boolean getTransformation(long currentTime, GLObject object) {
        boolean result = super.getTransformation(currentTime, object);
        for (GLAnimation animation : mAnimations) {
            animation.getTransformation(currentTime, object);
        }
        return result;
    }

    @Override
    public void setFillAfter(boolean fillAfter) {
        super.setFillAfter(fillAfter);
        for (GLAnimation animation : mAnimations) {
            animation.setFillAfter(fillAfter);
        }
    }
}

File: app/src/main/java/com/benhero/glstudio/base/GLAnimationSet.kt
package com.benhero.glstudio.base

import java.util.ArrayList

/**
 * GL动画集合
 *
 * @author Benhero
 */
class GLAnimationSet : GLAnimation() {
    private val mAnimations = ArrayList<GLAnimation>()

    val animations: List<GLAnimation>
        get() = mAnimations

    override var startTime: Long
        get() = super.startTime
        set(startTime) {
            super.startTime = startTime
            for (animation in mAnimations) {
                animation.startTime = startTime
            }
        }

    override var duration: Long
        get() = super.duration
        set(duration) {
            super.duration = duration
            for (animation in mAnimations) {
                animation.duration = duration
            }
        }

    override var isFillAfter: Boolean
        get() = super.isFillAfter
        set(fillAfter) {
            super.isFillAfter = fillAfter
            for (animation in mAnimations) {
                animation.isFillAfter = fillAfter
            }
        }

    fun addAnimation(animation: GLAnimation) {
        mAnimations.add(animation)
    }

    /**
     * 根据子动画更新动画集合的时间
     */
    fun updateTimeByChildren() {
        var startTime: Long = 0
        var endTime: Long = 0
        for (animation in mAnimations) {
            startTime = Math.min(mStartTime, animation.mStartTime)
            endTime = Math.max(animation.startTime + animation.duration, endTime)
        }
        mStartTime = startTime
        mDuration = endTime - startTime
    }

    override fun getTransformation(currentTime: Long, `object`: GLObject): Boolean {
        val result = super.getTransformation(currentTime, `object`)
        for (animation in mAnimations) {
            animation.getTransformation(currentTime, `object`)
        }
        return result
    }
}
